### PR TITLE
python39Packages.pytest-xdist: add hook to add flags

### PR DESCRIPTION
**Pull request by Artturin**
many of our packages have pytest-xdist in their inputs however
oftentimes the required flags aren't added to pytest flags
since we dont use tox and other frameworks

some examples of improvements
(my pc has ryzen 9 5900x)
hypothesis 11m:15s on hydra -> <1m on my pc
hypothesmith 8m on my pc -> <1m on my pc

the whole graph from pytest-xdist (tested by adding postPatch with true to xdist)
time nix build ".#python3Packages.hypothesmith"
19m:57s -> 9m:13s

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/pytest-xdist/setup-hook.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+pytestXdistPhase() {
+    pytestFlagsArray+=("-n$NIX_BUILD_CORES" "--forked" )
+}
+if [ -z "${dontUsePytestXdist-}" ] && [ -z "${pytestCheckPhase-}" ]; then
```
[the remark this answers is not in the document]
Comment (Artturin) [reply pkgs/development/python-modules/pytest-xdist/setup-hook.sh]: fixed, thanks for the review.

**Comment by FRidh** [comment]
Not relevant right now, but in case `checkInputs` gets split into a `checkNativeInputs` and `checkInputs` we may need to have this library in both in order to have the hook run. Though I am not sure about that.

**Review by FRidh — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]
I think this will be very valuable. Thank you!

File: pkgs/development/python-modules/pytest-xdist/default.nix [hunk 2]
```diff
@@ -25,9 +26,11 @@ buildPythonPackage rec {
   buildInputs = [
     pytest
   ];
-  checkInputs = [ pytestCheckHook filelock ];
+  checkInputs = [ pytestCheckHook filelock pexpect ];
   propagatedBuildInputs = [ execnet pytest-forked psutil ];
 
+  pytestFlagsArray = [ "-n $NIX_BUILD_CORES" "--forked" ]; # pytest can already use xdist at this point
```
Comment (roberth) [inline pkgs/development/python-modules/pytest-xdist/default.nix:32]: Doesn't this duplicate what happens in the setup hook?

Comment (Artturin) [reply pkgs/development/python-modules/pytest-xdist/default.nix:32]: i don't think the setup hook is used at this point yet?

Comment (mweinelt) [inline pkgs/development/python-modules/pytest-xdist/default.nix:32]: I do prefer longopts in non-interactive usage and ideally one argument per line for better diffs.

```suggestion
  pytestFlagsArray = [
    # pytest can already use xdist at this point
    "--numprocesses $NIX_BUILD_CORES"
    "--forked"
  ];
```

The `--forked` argument is apparently on the way out: https://github.com/pytest-dev/pytest-forked

File: pkgs/development/python-modules/pytest-xdist/default.nix [hunk 3]
```diff
@@ -25,9 +26,11 @@ buildPythonPackage rec {
   buildInputs = [
     pytest
   ];
-  checkInputs = [ pytestCheckHook filelock ];
+  checkInputs = [ pytestCheckHook filelock pexpect ];
```
Comment (mweinelt) [inline pkgs/development/python-modules/pytest-xdist/default.nix:29]: Where does pexpect come into play?

**Comment by mweinelt** [comment]
I think this is a reasonable change and we can test it in #166489 if you rebase onto that.

File: pkgs/development/python-modules/pytest-xdist/default.nix (hunk 2 above)
Comment (Artturin) [reply pkgs/development/python-modules/pytest-xdist/default.nix:32]: just tested and yep its not used at this point



if i apply the following the build fails with
`python3.9-pytest-xdist> substitute(): ERROR: file '/nix/store/q4qj7q82vrxfwnmymyhdwxid0sr0rjnn-hook' does not exist`

```patch
diff --git a/pkgs/development/python-modules/pytest-xdist/default.nix b/pkgs/development/python-modules/pytest-xdist/default.nix
index 75fd066c76d..ae23bf3f31b 100644
--- a/pkgs/development/python-modules/pytest-xdist/default.nix
+++ b/pkgs/development/python-modules/pytest-xdist/default.nix
@@ -1,4 +1,5 @@
 { lib
+, makeSetupHook
 , buildPythonPackage
 , fetchPypi
 , pythonOlder
@@ -12,6 +13,10 @@
 , pexpect
 }:
 
+let
+  setupHook' = makeSetupHook {} ./setup-hook.sh;
+in
+
 buildPythonPackage rec {
   pname = "pytest-xdist";
   version = "2.5.0";
@@ -26,11 +31,9 @@ buildPythonPackage rec {
   buildInputs = [
     pytest
   ];
-  checkInputs = [ pytestCheckHook filelock pexpect ];
+  checkInputs = [ pytestCheckHook filelock pexpect setupHook' ];
   propagatedBuildInputs = [ execnet pytest-forked psutil ];
 
-  pytestFlagsArray = [ "-n $NIX_BUILD_CORES" "--forked" ]; # pytest can already use xdist at this point
-
   # access file system
   disabledTests = [
     "test_distribution_rsyncdirs_example"
@@ -43,7 +46,7 @@ buildPythonPackage rec {
     "test_internal_errors_propagate_to_controller"
   ];
 
-  setupHook = ./setup-hook.sh;
+  setupHook = setupHook';
 
   meta = with lib; {
     description = "Pytest xdist plugin for distributed testing and loop-on-failing modes";
```

**Review by mweinelt — APPROVED** [review]

**Comment by roberth** [comment]
I don't usually package python, so I'll unsubscribe after my superficial observation earlier.

**Pull request merged**